xpti/include/xpti/xpti_trace_framework.hpp

```diff
@@ -479,84 +559,930 @@ inline std::string readMetadata(const metadata_t::value_type &MD) {
+  /// and adds the desired metadata.
+  /// @return A reference to this tracepoint_scope_t object.
+  ///
+  tracepoint_scope_t &
+  addMetadata(const std::function<void(xpti::trace_event_data_t *)> &Callback) {
+    if (xptiCheckTraceEnabled(MStreamId, MTraceType) && MTraceEvent) {
+      Callback(MTraceEvent);
+    }
+    return *this;
+  }
+
+  bool isTraceTypeScoped() {
+    if (MTraceType != (uint16_t)xpti::trace_point_type_t::signal &&
+        MTraceType != (uint16_t)xpti::trace_point_type_t::graph_create &&
+        MTraceType != (uint16_t)xpti::trace_point_type_t::node_create &&
+        MTraceType != (uint16_t)xpti::trace_point_type_t::edge_create &&
+        MTraceType != (uint16_t)xpti::trace_point_type_t::queue_create &&
+        MTraceType != (uint16_t)xpti::trace_point_type_t::queue_destroy &&
+        MTraceType != (uint16_t)xpti::trace_point_type_t::diagnostics)
+      return true;
+    else
+      return false;
+  }
+
 private:
-  xpti::trace_event_data_t *m_object, *m_parent;
-  uint8_t m_stream_id;
-  uint16_t m_trace_type;
-  const void *m_user_data;
-  uint64_t m_instance;
+  void selfNotifyEnd() {
+    // selfNotify() is supposed to capture the scope of all the tracepoints we
+    // have enbled; if we need to limit the notifications only to when the trace
+    // scope has valid TLS data, we can add the follwoing IF block if
+    // (!m_data.isValid())
+    //   return;
+    uint16_t traceType = (uint16_t)xpti::trace_point_type_t::function_end;
+    if (MSelfNotify && MDefaultStreamId &&
+        xptiCheckTraceEnabled(MDefaultStreamId, traceType)) {
+      const void *UserData = static_cast<const void *>(MCallerFuncName);
+      xptiNotifySubscribers(MDefaultStreamId, traceType, nullptr, MTraceEvent,
+                            MCorrelationId, UserData);
+    }
+  }
+
+  void selfNotifyBegin() {
+    // selfNotify() is supposed to capture the scope of all the tracepoints we
+    // have enbled; if we need to limit the notifications only to when the trace
+    // scope has valid TLS data, we can add the follwoing IF block if
+    // (!m_data.isValid())
+    //   return;
+    uint16_t traceType = (uint16_t)xpti::trace_point_type_t::function_begin;
+    if (MSelfNotify && MDefaultStreamId &&
+        xptiCheckTraceEnabled(MDefaultStreamId, traceType)) {
+      MCorrelationId = xptiGetUniqueId();
+      const void *UserData = static_cast<const void *>(MCallerFuncName);
+      xptiNotifySubscribers(MDefaultStreamId, traceType, nullptr, MTraceEvent,
+                            MCorrelationId, UserData);
+    }
+  }
+
+  /// Indicates if this is the top-level tracepoint
+  bool MTop = false;
+  /// Indicates if the tracepoint should notify the scope of the tracepoint
+  bool MSelfNotify = true;
+  /// Indicates if scoped notification is enabled
+  bool MScopedNotify = false;
+  /// Stores the tracepoint data from creating the tracepoint or from
+  /// inheritance
+  xpti_tracepoint_t *MData;
```

Comment:
needs initialization, not all of these members is covered in constructor. code analyzer will complain about this

Reply:
Fixed